primitives/keystore/src/lib.rs

```diff
@@ -174,6 +176,89 @@ pub trait Keystore: Send + Sync {
 
+	/// Returns all the bandersnatch public keys for the given key type.
+	#[cfg(feature = "bandersnatch-experimental")]
+	fn bandersnatch_public_keys(&self, key_type: KeyTypeId) -> Vec<bandersnatch::Public>;
+
+	/// Generate a new bandersnatch key pair for the given key type and an optional seed.
+	///
+	/// Returns an `bandersnatch::Public` key of the generated key pair or an `Err` if
+	/// something failed during key generation.
+	#[cfg(feature = "bandersnatch-experimental")]
+	fn bandersnatch_generate_new(
+		&self,
+		key_type: KeyTypeId,
+		seed: Option<&str>,
+	) -> Result<bandersnatch::Public, Error>;
+
+	/// Generate an bandersnatch signature for a given message.
+	///
+	/// Receives [`KeyTypeId`] and an [`bandersnatch::Public`] key to be able to map
+	/// them to a private key that exists in the keystore.
+	///
+	/// Returns an [`bandersnatch::Signature`] or `None` in case the given `key_type`
+	/// and `public` combination doesn't exist in the keystore.
+	/// An `Err` will be returned if generating the signature itself failed.
+	#[cfg(feature = "bandersnatch-experimental")]
+	fn bandersnatch_sign(
+		&self,
+		key_type: KeyTypeId,
+		public: &bandersnatch::Public,
+		msg: &[u8],
+	) -> Result<Option<bandersnatch::Signature>, Error>;
+
+	/// Generate a bandersnatch VRF signature for the given data.
+	///
+	/// Receives [`KeyTypeId`] and an [`bandersnatch::Public`] key to be able to map
+	/// them to a private key that exists in the keystore.
+	///
+	/// Returns `None` if the given `key_type` and `public` combination doesn't
+	/// exist in the keystore or an `Err` when something failed.
+	#[cfg(feature = "bandersnatch-experimental")]
+	fn bandersnatch_vrf_sign(
+		&self,
+		key_type: KeyTypeId,
+		public: &bandersnatch::Public,
+		input: &bandersnatch::vrf::VrfSignData,
+	) -> Result<Option<bandersnatch::vrf::VrfSignature>, Error>;
+
+	/// Generate a bandersnatch VRF (pre)output for a given input data.
+	///
+	/// Receives [`KeyTypeId`] and an [`bandersnatch::Public`] key to be able to map
+	/// them to a private key that exists in the keystore.
+	///
+	/// Returns `None` if the given `key_type` and `public` combination doesn't
+	/// exist in the keystore or an `Err` when something failed.
+	#[cfg(feature = "bandersnatch-experimental")]
+	fn bandersnatch_vrf_output(
+		&self,
+		key_type: KeyTypeId,
+		public: &bandersnatch::Public,
+		input: &bandersnatch::vrf::VrfInput,
+	) -> Result<Option<bandersnatch::vrf::VrfOutput>, Error>;
+
+	/// Generate a bandersnatch ring-VRF signature for the given data.
+	///
+	/// Receives [`KeyTypeId`] and an [`bandersnatch::Public`] key to be able to map
+	/// them to a private key that exists in the keystore.
+	///
+	/// Also takes a [`bandersnatch::ring_vrf::RingProver`] instance obtained from
+	/// a valid [`bandersnatch::ring_vrf::RingContext`].
+	///
+	/// The signature is valid if the signing key is part of the ring from which
+	/// the `RingProver` has been derived.
```

Comment:
If a signature is produced with a key whose corresponding public key doesn't belong to the ring (aka a group of public keys) then a signature will be produced anyway. However the signature is useless as it can't be verified using the ring verifier.

I added some docs and a test clarifying this in https://github.com/paritytech/substrate/pull/14412/commits/2b2676b14263d75c3c9412197dc0fb51b48dfffa